auth/auth.go

```diff
@@ -0,0 +1,174 @@
+// that auth is providing.
+const ClientTimeout time.Duration = time.Second * 10
+
+const (
+	defaultDeviceDesc string = "desktop-windows"
+	deviceTokenURL    string = "https://my.remarkable.com/token/json/2/device/new"
+	userTokenURL      string = "https://my.remarkable.com/token/json/2/user/new"
+)
+
+var defaultTokenStore FileTokenStore
+
+// Auth is a structure containing authentication gears to fetch and hold tokens
+// for interacting authenticated with the Remarkable Cloud API.
+type Auth struct {
+	ts TokenStore
+
+	// Refresh can be used to force a refresh of the UserToken.
+	Refresh bool
+}
+
+func New() *Auth {
+	return NewFromStore(&defaultTokenStore)
+}
+
+func NewFromStore(ts TokenStore) *Auth {
+	return &Auth{ts, false}
+}
+
+// RegisterDevice will make an HTTP call to the Remarkable API using the provided code
+// to register a new device. The code should be gathered at https://my.remarkable.com/generator-device.
+// The DeviceToken is then attached to the Auth instance.
+func (a *Auth) RegisterDevice(code string) error {
+	uuid, err := uuid.NewV4()
+	if err != nil {
+		return err
+	}
+
+	body, err := json.Marshal(map[string]string{
+		"code":       code,
+		"deviceDesc": defaultDeviceDesc,
+		"deviceID":   uuid.String(),
+	})
+
+	req, err := http.NewRequest("POST", deviceTokenURL, bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+
+	client := &http.Client{}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return errors.New("auth: can't register device")
+	}
+
+	bearer, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	tks := TokenSet{
+		DeviceToken: string(bearer),
+		UserToken:   "",
+	}
+
+	// persist device token and reset user token
+	if err := a.ts.Save(tks); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// renewToken will try to fetch a userToken from a deviceToken.
+func renewToken(deviceToken string) (userToken string, err error) {
+	req, err := http.NewRequest("POST", userTokenURL, nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set("Authorization", "Bearer "+deviceToken)
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("auth: can't renew token (HTTP %d)", resp.StatusCode)
+	}
+
+	bearer, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return string(bearer), nil
+}
+
+// Token will return a UserToken fetching it before if nil.
+func (a *Auth) Token() (string, error) {
+	tks, err := a.ts.Load()
+	if err != nil {
+		return "", err
+	}
+
+	if tks.UserToken != "" && !a.Refresh {
+		return tks.UserToken, nil
+	}
+
+	if tks.DeviceToken == "" {
+		return "", errors.New("auth: nil DeviceToken, please register device")
+	}
+
+	tks.UserToken, err = renewToken(tks.DeviceToken)
```

Comment:
Just notice that I could reset the `a.Refresh` flag to `false` when the token is renewed.